### PR TITLE
HDDS-6523. Refine OMClientRequest Classes to handle bucket layout better.

### PR DESCRIPTION
**Pull request by JyotinderSingh**
## What changes were proposed in this pull request?

This Jira improves instantiation logic for OMClientRequest classes and removes duplicated validation logic across different classes.
We now maintain a mapping of different request types to their associated classes based on bucket layout. This should allow for stronger validations - making sure we are operating on buckets with the correct request class flavors (FSO, OBS).

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6523

## How was this patch tested?

Integration tests.

### CODE REVIEW

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMKeyRequestFactory.java [hunk 1]
```diff
@@ -66,43 +55,56 @@ public static OMKeyRequest createRequest(OMRequest omRequest,
...
     case DeleteKey:
       keyArgs = omRequest.getDeleteKeyRequest().getKeyArgs();
-      omKeyRequest =
-          OMKeyDeleteRequest.getInstance(keyArgs, omRequest, ozoneManager);
+      omKeyRequest = BucketLayoutAwareOMKeyRequestFactory.createRequest(
+          keyArgs.getVolumeName(), keyArgs.getBucketName(), omRequest,
+          ozoneManager);
       break;
     case DeleteKeys:
       return new OMKeysDeleteRequest(omRequest);
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMKeyRequestFactory.java:93]: @JyotinderSingh Please add this also to the factory class.

   ```
 case DeleteKeys:
      OzoneManagerProtocolProtos.DeleteKeyArgs deleteKeyArgs =
          omRequest.getDeleteKeysRequest().getDeleteKeys();
      omKeyRequest = BucketLayoutAwareOMKeyRequestFactory.createRequest(
          deleteKeyArgs.getVolumeName(), deleteKeyArgs.getBucketName(), omRequest,
          ozoneManager);
      break;
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMKeyRequestFactory.java [hunk 2]
```diff
@@ -66,43 +55,56 @@ public static OMKeyRequest createRequest(OMRequest omRequest,
     switch (cmdType) {
     case CreateDirectory:
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMKeyRequestFactory.java:56]: As per the online discussion with @JyotinderSingh the idea came to remove the OmKeyReqFactory and can replace the check like below:

 ```
   case CreateDirectory:
      keyArgs = omRequest.getCreateDirectoryRequest().getKeyArgs();
      volName = keyArgs.getVolumeName();
      buckName = keyArgs.getBucketName();
      break;
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMKeyRequestFactory.java (hunk 1 above)
Comment (JyotinderSingh) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMKeyRequestFactory.java]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/BucketLayoutAwareOMKeyRequestFactory.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+        OMKeyDeleteRequest.class,
+        BucketLayout.OBJECT_STORE);
+    addRequestClass(Type.DeleteKey,
+        OMKeyDeleteRequestWithFSO.class,
+        BucketLayout.FILE_SYSTEM_OPTIMIZED);
+
+    // DeleteKeys
+    addRequestClass(Type.DeleteKeys,
+        OMKeysDeleteRequest.class,
+        BucketLayout.OBJECT_STORE);
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/BucketLayoutAwareOMKeyRequestFactory.java:134]: @JyotinderSingh Please add `OMKeysDeleteRequestWithFSO.class` to this set.

Comment (JyotinderSingh) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/BucketLayoutAwareOMKeyRequestFactory.java:120]: Done

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestBucketLayoutAwareOMKeyFactory.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      LoggerFactory.getLogger(TestBucketLayoutAwareOMKeyFactory.class);
+
+  /**
+   * Validates instantiation of each OMKeyRequest present inside the
+   * (CmdType + BucketLayout) -> RequestClass Mapping.
+   */
+  @Test
+  public void testGetRequestInstanceFromMap() {
+    // Iterate over each OMKeyRequest present in the mapping.
+    BucketLayoutAwareOMKeyRequestFactory.OM_KEY_REQUEST_CLASSES.forEach(
```
Comment (rakeshadr) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestBucketLayoutAwareOMKeyFactory.java:45]: Please do more assertions like below:

```
    List<OMKeyRequest> omKeyReqsFSO = new ArrayList<>();
    List<OMKeyRequest> omKeyReqsLegacy = new ArrayList<>();
    List<OMKeyRequest> omKeyReqsOBS = new ArrayList<>();
```
....
....
```
         OMKeyRequest omKeyRequest = BucketLayoutAwareOMKeyRequestFactory
                  .getRequestInstanceFromMap(
                      getDummyOMRequest(), k,
                      BucketLayout.FILE_SYSTEM_OPTIMIZED);
              Assert.assertEquals(BucketLayout.FILE_SYSTEM_OPTIMIZED, omKeyRequest.getBucketLayout());
              omKeyReqsFSO.add(omKeyRequest);
```
.....
....
....

```
              OMKeyRequest omKeyRequest1 = BucketLayoutAwareOMKeyRequestFactory
                  .getRequestInstanceFromMap(
                      getDummyOMRequest(), k, BucketLayout.LEGACY);
              Assert.assertEquals(BucketLayout.LEGACY, omKeyRequest1.getBucketLayout());
              omKeyReqsLegacy.add(omKeyRequest1);

              OMKeyRequest omKeyRequest2 = BucketLayoutAwareOMKeyRequestFactory
                  .getRequestInstanceFromMap(
                      getDummyOMRequest(), k, BucketLayout.OBJECT_STORE);
              Assert.assertEquals(BucketLayout.OBJECT_STORE, omKeyRequest2.getBucketLayout());
              omKeyReqsOBS.add(omKeyRequest2);
```
......
......


```
    Assert.assertEquals(12, omKeyReqsFSO.size());
    Assert.assertEquals(13, omKeyReqsLegacy.size());
    Assert.assertEquals(13, omKeyReqsOBS.size());
```

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/BucketLayoutAwareOMKeyRequestFactory.java [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+   */
+  public static OMKeyRequest createRequest(String volumeName, String bucketName,
+                                           OMRequest omRequest,
+                                           OzoneManager ozoneManager)
+
+      throws IOException {
+    // Get the bucket layout of the bucket being accessed by this request.
+    // While doing this we make sure we are resolving the real bucket in case of
+    // link buckets.
+    BucketLayout bucketLayout =
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/BucketLayoutAwareOMKeyRequestFactory.java:219]: done

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestBucketLayoutAwareOMKeyFactory.java (hunk 4 above)
Comment (JyotinderSingh) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestBucketLayoutAwareOMKeyFactory.java]: Done

**Comment by JyotinderSingh** [comment]
Thank you for the reviews @rakeshadr, I have addressed your comments in the latest patch. Please let me know if it looks good to you?

**Review by rakeshadr — APPROVED** [review]
+1LGTM , thanks @JyotinderSingh for the continuous efforts in refining the layout code base!

**Pull request merged**